### PR TITLE
MIR EndRegion Statements (was MIR dataflow for Borrows)

### PR DESCRIPTION
**Pull request by pnkfelix**
This PR adds an `EndRegion` statement to MIR (where the `EndRegion` statement is what terminates a borrow).

An earlier version of the PR implemented a dataflow analysis on borrow expressions, but I am now factoring that into a follow-up PR so that reviewing this one is easier. (And also because there are some revisions I want to make to that dataflow code, but I want this PR to get out of WIP status...)

This is a baby step towards MIR borrowck. I just want to get the review process going while I independently work on the remaining steps.

### CODE REVIEW

**Comment by eddyb** [comment]
So what happened to the plans for keeping this information in a side table? cc @nikomatsakis

File: src/librustc_mir/build/scope.rs
Comment (nagisa) [inline src/librustc_mir/build/scope.rs]: Doesn’t early exit from the scope (i.e. `exit_scope` down below) need a similar treatment?

**Comment by nagisa** [comment]
> So what happened to the plans for keeping this information in a side table?

It makes sense to me to keep these inline, even if just because we already keep StorageLiveDead inline.

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: You're right, I had overlooked that. Will fix.

(I had been hoping to avoid attempting to write unit tests for this, but maybe I need to do so.)

**Comment by nikomatsakis** [comment]
@eddyb 

> So what happened to the plans for keeping this information in a side table?

Well, I think I turned sour on this idea because it's hard to keep such a thing 'in sync' if the MIR is changed in any way. It's very fragile. Also, I think these instructions are temporary (once we adopt the NLL strategy, we'll be able to drop them) and there shoudl be relatively few (you only need them for regions that appear in a loan, though I'm not sure if @pnkfelix handles that since I've not read the PR yet). But it is I suppose an option.

**Comment by pnkfelix** [comment]
@arielb1 has convinced me that I need to add some code-gen tests, even if their initial state is based solely on eyeballing the output to confirm that it is sane.

File: src/librustc_borrowck/borrowck/mir/dataflow/impls.rs
Comment (arielb1) [inline src/librustc_borrowck/borrowck/mir/dataflow/impls.rs]: Use the `new_index` macro?

Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/mir/dataflow/impls.rs]: ah I forgot about that; will fix.

**Comment by pnkfelix** [comment]
Bisection reports that commit 039f5f95ea6646741f9fdd7d6cb9335366d5c82b injected a bug... investigating

**Comment by pnkfelix** [comment]
@arielb1 has suggested that I had a static check that no regions reach mir `return` nor `resume` aka a "no leak" check. This is a good idea; it remains to be implemented.

**Comment by pnkfelix** [comment]
(Testing reveals that I missed EndRegion's on the unwind path ... fixing in tandem with adding the above suggested static check.)

File: src/librustc/util/ppaux.rs
Comment (nikomatsakis) [inline src/librustc/util/ppaux.rs]: does `-Zverbose` not do it for you? Too verbose?

Comment (pnkfelix) [inline src/librustc/util/ppaux.rs]: Yes, `-Zverbose` is far too verbose, especially when it is embedded within a `&<region> <lvalue>` borrow-expression.

I don't want or care about the internal structure of the region value for these cases; it is much more convenient to look at the unique id's being generated here (by taking advantage of the fact that one cannot write a lifetime of the form `'<number><alphanum>`

**Comment by pnkfelix** [comment]
Okay! Finally green!

I think this is ready for review now; I took out the dataflow so that this PR is focused solely on the EndRegion additions, since those ended up being more complex than I had anticipated. I'll have a followup PR with dataflow.

File: src/librustc_mir/build/expr/as_rvalue.rs
Comment (arielb1) [inline src/librustc_mir/build/expr/as_rvalue.rs]: Is this also correct for things like `return`? As in:
```rust
'a: loop {
    loop {
        if second_iter {
            break 'a; // want to generate `EndRegion('a)` here
        } else {
            &'a x
        }
    }
}
```

File: src/test/mir-opt/end_region_2.rs
Comment (arielb1) [inline src/test/mir-opt/end_region_2.rs]: Who does `RegionEnd('17ce)` here?

Comment (pnkfelix) [inline src/test/mir-opt/end_region_2.rs]: Oops, these tests make it easy to omit statements. There should be a `EndRegion('17ce)` in `bb2` between lines 49 and 50 here, I think. (At least that's what I'm getting from my local build of this branch.)

Comment (pnkfelix) [inline src/test/mir-opt/end_region_2.rs]: (good catch!)

File: src/librustc_mir/build/expr/as_rvalue.rs
Comment (pnkfelix) [inline src/librustc_mir/build/expr/as_rvalue.rs]: I don't mind trying to do it that way. I can gather data on how many extraneous `EndRegion` statements are emitted and then removed.

Comment (arielb1) [inline src/librustc_mir/build/expr/as_rvalue.rs]: That sounds like it could be cleaner. Could you try it?

**Comment by arielb1** [comment]
Sorry for being so late in review.

File: src/librustc_mir/build/expr/as_rvalue.rs
Comment (pnkfelix) [inline src/librustc_mir/build/expr/as_rvalue.rs]: Will do!

**Comment by shepmaster** [comment]
Looks like the ball is back in your court, @pnkfelix! Do you believe you'll be able to address the suggestions and merge conflicts?

**Comment by shepmaster** [comment]
Thanks for the PR, but we are going to close this due to inactivity and to keep the queues clear. Please reopen when you've been able to address the feedback!

**Comment by pnkfelix** [comment]
Okay so I finally got around to gathering the data promised above on the effectiveness of pre-filtering `EndRegion` statements by only emitting them for regions that were borrowed.

I'll be posting a comment shortly that summarizes my findings. The data as I see it has mostly reinforced my belief that the pre-filter does pay for itself, but I want to share the data before I say more.

**Comment by pnkfelix** [comment]
The executive summary is that if you do not prefilter the `EndRegion` statements, the number of statements blows up by a factor between 1.4x and 1.8x compared to if you *do* prefilter them.

Out of convenience, I used the crates of the compiler itself as the basis for my analysis.

Here is a gist with the script I used, and the two sets of output from the two runs of the script with and without the prefilter: https://gist.github.com/pnkfelix/e9131dafb9ad3e6b87c29c5ba3c998e1

And here is a spreadsheet summarizing the data: https://docs.google.com/spreadsheets/d/1n8PY2H-ogeGFR_-WKsZi-wFwzduvcSR84vXnz7dw2OI/edit?usp=sharing

As a typical example, the `rustc` crate emits 452,556 `EndRegion` statements without the prefilter, and 66,187 `EndRegion` statements with the prefilter. The number of statements with the filter (i.e. the supposed minimum once `EndRegion` support lands) for that crate is 698,429. So, without the filter there's about a 1.5x blowup for that crate (1.5 ≈ (698,429 + 452,556)/698,429).

----

For those readers who want to understand the source of this blowup:
 * Without a filter, the number of `EndRegion` statements generated is equal to the number of lexical scopes (i.e. code-extents) in the HAIR. With a filter, the number of `EndRegion` statements is equal to the number of scopes that have actually been involved in a *borrow* (e.g. a `&<scope> <expr>` expression, or `ref` pattern, or an auto-borrow in a method call or field dereference).
 * As a back of the envelope calculation, we can upper-bound the number of such scopes by simply counting the number of borrows.
 * The point is: there tend to be so many more `EndRegion` statements that can be trivially filtered is that there usually *far* more lexical scopes than there are borrows in any given program.

----

There are three hypothetical reasons I can identify that one might object to the prefilter:

 1. Hashset memory use: Its not guaranteed to always use strictly less memory than a non-prefiltered variant, since it maintains `seen-borrows`, a hashset of code-extents that it has seen in some borrow. But the amount of memory used by that hashset should be O(min(#borrows, #code-extents)). 
   * From the data, I note again that the number of borrows tends to be much smaller than the number of code-extents. Thus I conclude that the cost of `seen-borrows` should be far less than the cost of no prefilter.
   * (Skimming over the data, I'd say that the number of extraneous `EndRegion` statements will often be about 5x to 8x times the number of entries in the hashset. And that's *before* you consider that an `EndRegion` statement is at least 192 bytes currently, while a `CodeExtent` is a `u32`, which makes the relative cost even more pronounced, even allowing for a fair amount of excess capacity in the hashset)

 2. Correctness of Prefilter. We must correctly construct `seen-borrows`, or else we risk missing an `EndRegion` that we were supposed to emit. However, a bug like this should at worst result in borrowck rejecting code that it should accept; i.e., the risk associated with forgetting to emit an `EndRegion` is less scary to me than the risk associated with over-eagerly emitting an `EndRegion` in an unsound manner (something that is not a problem right now, but will be of more interest when we actually start implementing non-lexical-lifetimes in earnest).

 3. Complexity of Prefilter: adding this code has a maintenance cost. (There's no pat argument against this point.)

So, I think I've discharged the first two hypothetical objections. The third one, about the maintenance cost, leads us to a value judgement: Should we strive to keep the MIR construction as simple as possible, at a cost of 1.4x to 1.8x increase in statement count in the initially constructed MIR? Or is that cost high enough to justify a prefilter?

**Comment by pnkfelix** [comment]
Note: I still need to address the other feedback on the PR. In particular I need to double-check out how my code behaves on this example that @arielb1 provided: 

```rust
'a: loop {
    loop {
        if second_iter {
            break 'a; // want to generate `EndRegion('a)` here
        } else {
            &'a x
        }
    }
}
```

**Comment by nikomatsakis** [comment]
Some thoughts:

I agree that 1.5x-2x is bad but not **as** bad as I feared.

With respect to peak memory usage, one interesting thing is that -- on the latest master -- we now process MIR "one at a time". In other words, we don't build the MIR for all functions and *then* simplify them, instead we build the MIR for one function and then simplify it etc before starting in on the next one. This should help reduce peak memory usage.

We have had persistent problems with very large constants -- though I think these tend to be problems not of peak memory usage **per se** but specifically **stack usage**. Does this choice affect stack usage in any way? (I wouldn't think so.)

Naturally, we don't expect this "EndRegion" stuff to last very long (e.g., I'd like to have NLL in place before the end of year). I'm not sure which side that tilts for -- you can view it either way. Probably that time-scale is long enough that it doesn't really matter that much. 

I guess a key question is *how complex* the addition of the filter is. It's not just that there's a hashmap, iirc, there is also some amount of complexity in terms of making sure we also have a place to insert the `EndRegion`, right? But I think that @pnkfelix chose to take a relatively naive path there. I'll have to look over the PR's current state, though, I'm mostly going on what he and I said when talking.

**Comment by arielb1** [comment]
@pnkfelix

Does the filter have to be flow-sensitive? I think a flow-insensitive filter (aka collect all lifetimes of ref/autorefs in a MIR in a big HashSet and use these) would be simpler and get us most of the way.

**Comment by arielb1** [comment]
@pnkfelix

Does the filter have to be flow-sensitive? I think a flow-insensitive filter (aka collect all lifetimes of ref/autorefs in a MIR in a big HashSet and use these) would be simpler and get us most of the way.

**Comment by Mark-Simulacrum** [comment]
@pnkfelix Looks like this is waiting on a response from you and there are some merge conflicts.

**Comment by pnkfelix** [comment]
@arielb1 I don't think my filter was ever flow-sensitive (at least not intentionally so?)

**Comment by arielb1** [comment]
@pnkfelix

> @arielb1 I don't think my filter was ever flow-sensitive (at least not intentionally so?)

It adds borrows to the set when it sees them, so you have cases such as this:
```Rust
'a: {
    let x;
    if cond() {
        return // should we generate EndRegion('a) here?
    }
    x = &'a y;
}
```

Of course, that can easily get us in trouble in the presence of back-edges (see my other example).

**Comment by pnkfelix** [comment]
@arielb1 ah okay. Yes I now see what you are getting at. My previous attempts to gather a hashset of code-extents prior to MIR construction did not go well, and I do not think I will enjoy attempting to inject EndRegions after MIR construction, so I will just kill the prefilter.

**Comment by arielb1** [comment]
@pnkfelix 

So we are just waiting for the "prefilter stability" concern + rebase.

**Comment by pnkfelix** [comment]
Okay, updated! Sorry that took so long; it was a real bear of a rebase for some reason.

I've switched from a prefilter to a post-filter. So we have an spike where the MIR contains a lot of unnecessary EndRegions that are then removed.

----

I thought more about your example today. Perhaps, if the memory usage is deemed excessive, we could put back in the pre-filter and *just* apply it to the EndRegion's that are generated when we naturally leave scopes via normal exits, while non-local exits like `break` and `return` would be unfiltered (i.e. they would generate a slew of EndRegions that would subsequently be removed by the postfilter as is done in the current PR.) At least, I *think* that more conservative prefilter should work, if I understand how MIR construction proceeds.

But in any case, I am hoping to just land this as is, and leave worrying about a pre-filter for some later work which might never be necessary.

File: src/librustc_mir/build/scope.rs
Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: So I'm not sure I understand what we are doing here. This is what I think is happening:

- We always emit `EndRegion` on the normal path, regardless of whether there has been a borrow.
- We emit `EndRegion` on the diverging path also, but only if there is a borrow with that region.

This seems inconsistent, so maybe I am missing something

File: src/librustc_mir/build/scope.rs
Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: maybe make this a closure to avoid creating a string when not emitting `debug!`?

File: src/librustc_mir/build/scope.rs
Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: I have no idea what's going on here :(

File: src/librustc_mir/build/scope.rs
Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: It would be really great to put a comment on this struct explaining the overall flow. e.g., when is `terminator_emitted` set to true, etc? Clearly, this is encoding some sort of state machine, but it's a somewhat opaque one.

Comment (nikomatsakis) [inline src/librustc_mir/build/scope.rs]: Also, surely `pub(crate)` is much broader than is needed... why not `pub(super)`? (Actually, just from idle searching, it appears that private would suffice, which would make me even happier.)

File: src/test/mir-opt/end_region_2.rs
Comment (nikomatsakis) [inline src/test/mir-opt/end_region_2.rs]: do we not expect `EndRegion('7_1rce)` on this path?

File: src/test/mir-opt/end_region_3.rs
Comment (nikomatsakis) [inline src/test/mir-opt/end_region_3.rs]: here we see the `EndRegion` I expected above

File: src/test/mir-opt/end_region_2.rs
Comment (nikomatsakis) [inline src/test/mir-opt/end_region_2.rs]: ok I see it comes later, nevermind -- but how did test pass before?!

**Review by nikomatsakis — APPROVED** [review]
r=me but with a few more comments about the overall structure, as indicated

File: src/librustc_borrowck/borrowck/mir/elaborate_drops.rs
Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/mir/elaborate_drops.rs]: Hmm, in an earlier version of this PR I definitely needed to address Resume in this loop, because otherwise it falls into that `patch.is_patched` assertion below. But I'm guessing that this was an artifact of the dataflow changes that have since been moved to a different PR. I'll take this out and save figuring out whether its still necessary for later.

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: considering how much `pub` proliferates in much of the compiler, I was figuring a `pub(crate)` was a conservative choice here...  anyway I'll look into priv.

Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: (`end_region_emitted` is accessed from `librustc_mir/build/expr/as_rvalue.rs`. But that's the only one that needs to be exposed I think.)

File: src/librustc_borrowck/borrowck/mir/elaborate_drops.rs
Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/mir/elaborate_drops.rs]: Nope, still necessary. Okay so I'll figure out what I did here that caused this to be exposed, and then improve the explanation in the comment.

Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/mir/elaborate_drops.rs]: My hypothesis is that something I did here is causing changes to compiler control flow overall so that the `patch.patch_map` now holds entries for one or more resume terminators (replacing the resumes with gotos). Before this change, we never exercised that aspect of the patching code, and so this iteration here would fall into the `_ => { ... }` block where it asserts `!self.patch.is_patched(bb)`. But in this case, the code *is* patched (with the aforementioned replacement).

That's my current hypothesis / reconstructed memory.

Working on confirming this, and figuring out a clearer way to explain it. (Not sure if its worth trying to avoid it, since the patch code itself does *seem* to support this scenario, even if its been untested...)

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: my goal had been to continue lazily constructing the diverge path.

Its easy to generate EndRegion statements on their own (if somewhat wasteful) and then delete them afterwards when they are trivial.

I don't know whether it is as easy to do something similar for a diverge path that has been unnecessarily introduced. And I don't think the kind of example that arielb1 showed (that led me to stop trying to prefilter `EndRegion` in general) exists for diverge paths.

Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: But also, re-reading your comment... I believe we will emit EndRegion on diverging paths even if there is no borrow, that one can observe *before* the post-filter.... maybe I should make some tests that illustrate this.

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: The `EndRegion` block is preallocated, it is the `block` field on the `DivergePathState`. When `build_diverge_scope` is called, it will make that `block` end with a terminator that jumps to the drops that are built by `build_diverge_scope`.

For the purposes of instrumentation, I wanted the debug info lines to indicate whether we only do this patching once, or if we repatch the same block many times. (It turns out that we do repatch the same block many times, at least in some cases.)

And, since the `fn terminate` method requires that the block not have any terminator previously installed, we need to first install `None` to sidestep that check.

So... its a big hack, but I'm not sure more comments will help the situation.

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: mmm there's two control flow paths that initialize `repatching`, and the two closure expressions would have distinct types. I've looked into using `&Fn() -> String` but then the borrowck says the closures are temporaries that do not live long enough. Are you intending that I go all the way to `Box<Fn() -> String>`? (Which does work, and I guess is slightly less intensive than formatting a string unconditionally, but its not as clear to be a significant win...)

File: src/test/mir-opt/end_region_2.rs
Comment (pnkfelix) [inline src/test/mir-opt/end_region_2.rs]: (We discussed on IRC, but the basic answer is that the mir-opt test infrastructure is fragile in that it only checks that the generated output is a supersequence of the provided input. In other words, if there are statements in the output that are not included in the input, that is not enough on its own to cause a test to fail. This makes some tests pass when you wouldn't expect.)

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: okay so now I've looked into it and I think I have a simpler construction technique. Will push to my branch soon.

File: src/librustc_borrowck/borrowck/mir/elaborate_drops.rs
Comment (pnkfelix) [inline src/librustc_borrowck/borrowck/mir/elaborate_drops.rs]: I'm still not very happy with the story here. I think I will probably revise the comment to try to reflect my lack of unsureness about how we got to this point.

File: src/librustc_mir/build/scope.rs
Comment (pnkfelix) [inline src/librustc_mir/build/scope.rs]: (simpler construction techinque is now up)

**Comment by Mark-Simulacrum** [comment]
```
[01:05:46] stderr:

[01:05:46] ------------------------------------------

[01:05:46] thread 'main' panicked at 'assertion failed: town_bike != my_bike', /checkout/src/test/run-pass/issue-3935.rs:21

[01:05:46] note: Run with `RUST_BACKTRACE=1` for a backtrace.

[01:05:46] 

[01:05:46] ------------------------------------------

[01:05:46] 

[01:05:46] thread '[run-pass] run-pass/issue-3935.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2472
```

**Comment by pnkfelix** [comment]
(i don't understand how that error can be occurring from this change.)

**Comment by alexcrichton** [comment]
Hm I think that could be an ARM-specific codegen bug?

**Comment by pnkfelix** [comment]
Same failure again. Amazing. I'm at a loss about how to narrow this down.

**Comment by pnkfelix** [comment]
I guess I have to assume that my addition of some no-op intermediate MIR basic blocks is tickling something somewhere, perhaps in LLVM.

**Comment by pnkfelix** [comment]
(I have been clued into how to use `docker` to attempt to reproduce the problem on a non-arm host; I am investigating.)

**Comment by pnkfelix** [comment]
(I have finally replicated the test failure via the `armhf-gnu` docker image. Now need to dig into what is causing it!)

**Comment by pnkfelix** [comment]
Narrowed problem down via `opt -opt-bisect-limit`. Seems like the execution problem is injected by the Tail Call Elimination pass. I still need to review the LLVM IR to try to understand whether this is an LLVM codegen bug (as I suspect) or if something I have done has caused the Tail Call Elimination pass to "legitimately" mis-optimize this code.

**Comment by nikomatsakis** [comment]
@pnkfelix nice hunting

**Comment by arielb1** [comment]
Looks like the bug is fixed by https://github.com/llvm-mirror/llvm/commit/f45bea4a62dbc32b4166497ab311646589e5b8f3. I'll have the fix in my Grand ARM Fix Rollup, which I should really get done Real Soon.

**Comment by pnkfelix** [comment]
@arielb1 helped me confirm that the LLVM bug occurs on our fork of LLVM, but not on LLVM's master copy.  And I have now confirmed that the nearest common ancestor between the two also exhibits the bug.

I am bisecting against the LLVM git history to try to isolate a patch for the bug that we might be able to apply to our fork.

**Comment by arielb1** [comment]
Given that the LLVM bug is fixed:
@bors r=nikomatsakis

**Pull request merged**
